File: src/main/java/org/project/heredoggy/domain/postgresql/dog/Dog.java
package org.project.heredoggy.domain.postgresql.dog;

import jakarta.persistence.*;
import lombok.*;
import org.project.heredoggy.domain.postgresql.shelter.shelter.Shelter;
import org.project.heredoggy.domain.postgresql.walk.reservation.UnavailableDate;
import org.springframework.data.annotation.CreatedDate;
import org.springframework.data.annotation.LastModifiedDate;
import org.springframework.data.jpa.domain.support.AuditingEntityListener;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;

@Entity
@Getter
@Setter
@AllArgsConstructor
@NoArgsConstructor
@Builder
@EntityListeners(AuditingEntityListener.class)
public class Dog {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    private String name;

    private int age;

    @Enumerated(EnumType.STRING)
    private Gender gender;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private BreedType breed;

    @Column(columnDefinition = "TEXT")
    private String personality; // 특이사항

    private Double weight;

    private Boolean isNeutered; // 중성화 여부

    private String foundLocation; // 발견 장소

    @Enumerated(EnumType.STRING)
    private DogStatus status = DogStatus.AVAILABLE; // 현재 상태(사용 가능)

    @CreatedDate
    @Column(name = "created_at")
    private LocalDateTime createdAt; // 등록 일시

    @LastModifiedDate
    @Column(name = "updated_at")
    private LocalDateTime updatedAt; // 수정 일자

    @Column(name = "view_count", nullable = false)
    private Long viewCount = 0L;

    @Builder.Default
    @OneToMany(mappedBy = "dog", cascade = CascadeType.ALL, orphanRemoval = true)
    private List<DogImage> images = new ArrayList<>();

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "shelter_id", nullable = false)
    private Shelter shelter;

    @OneToMany(mappedBy = "dog", cascade = CascadeType.ALL, orphanRemoval = true)
    private List<UnavailableDate> unavailableDates = new ArrayList<>();

    @OneToMany(mappedBy = "dog", cascade = CascadeType.ALL, orphanRemoval = true)
    private List<DogComment> comments = new ArrayList<>();
    public void addImage(DogImage image){
        images.add(image); // dog + dogImage 연관관계 설정
        image.setDog(this); // dogimage + dog 연관 설정 (양방향 설정)
    }
    public void increaseViewCount(){
        this.viewCount++;
    }
}

File: src/main/frontend/lib/models/dog.dart
class Dog {
  final int id;
  final String name;
  final int age;
  final String gender;
  final String personality;
  final double weight;
  final bool isNeutered;
  final String foundLocation;
  final List<String> imagesUrls;
  final String shelterName;

  Dog({
    required this.id,
    required this.name,
    required this.age,
    required this.gender,
    required this.personality,
    required this.weight,
    required this.isNeutered,
    required this.foundLocation,
    required this.imagesUrls,
    required this.shelterName,
  });

  factory Dog.fromJson(Map<String, dynamic> json) {
    return Dog(
      id: json['id'] is int ? json['id'] : int.parse(json['id'].toString()),
      name: json['name'] as String,
      age: json['age'] as int,
      gender: json['gender'] as String,
      personality: json['personality'] as String? ?? '',
      weight: (json['weight'] as num).toDouble(),
      isNeutered: json['isNeutered'] as bool,
      foundLocation: json['foundLocation'] as String,
      imagesUrls: List<String>.from(json['imagesUrls'] as List),
      shelterName: json['shelterName'] as String,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'name': name,
      'age': age,
      'gender': gender,
      'personality': personality,
      'weight': weight,
      'isNeutered': isNeutered,
      'foundLocation': foundLocation,
      'imagesUrls': imagesUrls,
      'shelterName': shelterName,
    };
  }
} 